src/content/blog/2024-04-10-the-telemetry-tale.md

```diff
@@ -0,0 +1,117 @@
+
+### Architecture and Instrumentation
+
+![The Rafiki telemetry architecture](/developers/img/blog/2024-04-10/telemetry-architecture.png)
+
+The [number of transactions](https://github.com/interledger/rafiki/blob/d3be6b8d151d8cebc32b862e52a7bb678674d48e/packages/backend/src/open_payments/payment/outgoing/lifecycle.ts#L84-L90) is extracted from the Open Payments outgoing payment lifecycle and the value metric is handled by a [telemetry middleware](https://github.com/interledger/rafiki/blob/d3be6b8d151d8cebc32b862e52a7bb678674d48e/packages/backend/src/payment-method/ilp/connector/core/middleware/telemetry.ts) layer inside the ILP connector core.
+
+In summary, before these metrics are sent to the Otel collectors we’ve implemented local privacy-preserving measures, including:
+- packetization which is inherent in the Interledger Protocol (ILP)
+- a currency conversion service that uses a custom AWS Lambda function that pulls daily exchange rates from configured sources and stores them in an Amazon S3 bucket
+- rounding data to a precision point determined by bucketing
+- local differential privacy in the form of adding Laplacian noise to anonymize transaction data
+
+Finally, we grappled with the dilemma of disentangling playground data from our test network and live production data. The [Test Network](https://github.com/interledger/testnet) (Testnet) is an open Interledger network using test money within example wallet and e-commerce applications. Rafiki also provides a [local playground](https://github.com/interledger/rafiki/tree/main/localenv) to experiment with. We’d like to monitor activity within the testing environments to gauge interest in the Interledger Protocol (ILP) and identify usage patterns by new participants. We’ve opted to provide users the choice between the test network (testnet) and the live network (livenet). To support this dual-environment approach, the infrastructure for telemetry was effectively expanded to include two separate services: one dedicated to collecting and managing telemetry data from the test network and another for the live network. This setup ensures that data from each environment is handled independently.
+
+We’re using an AWS ECS Fargate deployment for our Telemetry Cluster. The cluster holds services for our two environments: livenet and testnet, both of them having a Network Load Balancer (NLB) in order to load-balance over their own set of ECS task replicas. By using an NLB instead of the Application Load Balancer (ALB), we are not taking the hit of parsing the L7 ([OSI Model](https://en.wikipedia.org/wiki/OSI_model)) protocols and we do not have to do TLS termination at load balancer level. In our case, we just need to pass the data received to Prometheus, and application-layer routing decisions are not needed. Acting on L4 is of greater convenience, as all we need is the TCP packets being redirected to one of the running ECS tasks.
```

Comment:
```suggestion
Our telemetry cluster is deployed on AWS, utilizing the ECS Fargate service to support two distinct operational environments: the live network environment (livenet) and the test network environment (testnet). Since the two environments are separate, it is the Rafiki client that determines where to send the data. This way each environment has its own load balancer to handle incoming traffic and the load balancers themselves do not need to decrypt GRPC/HTTPS traffic to determine where it needs to go. This allows us to establish a TLS secure channel directly with the services behind the load balancers, rather than with a load balancer itself. This configuration optimizes our resource utilization, leading to a more efficient and cost-effective telemetry infrastructure. 
```